### PR TITLE
feat(website): add logging

### LINKED ISSUE #203: Website: add logging
We need some proper logging capabilities on the website.

I'd propose to do the following:
* Use a logging framework? (Loculus uses https://github.com/winstonjs/winston AFAIK).
* Add a logging endpoint so that we can get the client side logs to our server.

### PR DESCRIPTION
**Pull request by fengelniederhammer**
resolves #203

### Summary

* Using https://github.com/winstonjs/winston
* For client side code, there is now `/api/log` that accepts POST requests with log messages.

### CODE REVIEW

File: website/src/logger.ts [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+import DailyRotateFile from 'winston-daily-rotate-file';
+
+import { type InstanceLogger } from './types/logMessage.ts';
+
+let _logger: Logger | undefined;
+
+const getLogger = (): Logger => {
+    if (_logger === undefined) {
+        const logPath = import.meta.env.LOG_DIR;
+        const serverLogFile = `${logPath}/website.log`;
```
Comment (JonasKellerer) [inline website/src/logger.ts:11]:
```suggestion
        const serverLogFile = `${logPath}/dashboards_website.log`;
```
We have also other websites on the server. Just to make it more cleare where the log comes from.
Or we find a name for the backend and the website ;)

File: website/src/pages/api/log.ts [hunk 2]
```diff
@@ -0,0 +1,26 @@
+import type { APIRoute } from 'astro';
+
+import { getInstanceLogger } from '../../logger.ts';
+import { type InstanceLogger, logMessageSchema } from '../../types/logMessage.ts';
+
+export const POST: APIRoute = async ({ request }) => {
+    const body = await request.json();
```
Comment (JonasKellerer) [inline website/src/pages/api/log.ts:7]: Currently, everybody can access this endpoint. Should we restrict the access to only logged in users? This would prevent us from getting error messages from the organism pages. But in theory one can flood our endpoint his way?

File: website/src/types/ProblemDetail.ts [hunk 3]
```diff
@@ -2,11 +2,11 @@ import { z } from 'zod';
 
 export const problemDetailSchema = z
     .object({
-        type: z.string().url().optional(),
+        type: z.string().optional(),
         title: z.string().optional(),
         status: z.number().int().optional(),
         detail: z.string().optional(),
-        instance: z.string().url().optional(),
+        instance: z.string().optional(),
```
Comment (JonasKellerer) [inline website/src/types/ProblemDetail.ts:9]: From the docs it seems that instance and the type should be an uri. Why did you change it here?

File: website/.env.example [hunk 4]
```diff
@@ -1,6 +1,7 @@
 CONFIG_DIR=../backend/src/main/resources/
 BACKEND_URL=http://localhost:8080
 DASHBOARDS_ENVIRONMENT=dashboards-dev
+LOG_DIR=./logs
```
Comment (JonasKellerer) [inline website/.env.example:4]: I think we should also add the log level here, since we have no other place to document it.

File: website/src/middleware/loggingMiddleware.ts [hunk 5]
```diff
@@ -0,0 +1,13 @@
+import { defineMiddleware } from 'astro/middleware';
+
+import { getInstanceLogger } from '../logger.ts';
+
+const logger = getInstanceLogger('RequestLoggingMiddleware');
+
+export const loggingMiddleware = defineMiddleware(async (context, next) => {
+    const response = await next();
+    logger.info(
```
Comment (JonasKellerer) [inline website/src/middleware/loggingMiddleware.ts:9]: Do we really want to log every page load? I think it would be nice on debug.

**Review by JonasKellerer — COMMENTED** [review]
Nice work.
I think the only thing we need to discuss, is whether we want to log every request to our page. And if everybody should be able to access our /api/log endpoint.

File: website/src/types/ProblemDetail.ts (hunk 3 above)
Comment (fengelniederhammer) [reply website/src/types/ProblemDetail.ts:9]: It wasn't a URL in fact. 

Also this should be as resilient as possible. Here (on the website), it doesn't matter what kind of string it is. But it should not fail to parse objects, because then our error handling fails (which is bad!).

File: website/src/pages/api/log.ts (hunk 2 above)
Comment (fengelniederhammer) [reply website/src/pages/api/log.ts:7]: As discussed: 
* We want that every page can send log messages, also pages where users don't need to be logged in.
* There is no other proper protection, is it should be publicly available
* We only write log messages to the log.  We don't execute code or similar (assuming that winston doesn't have vulnerabilities). So we have a risk of being DDoSed / spamed, but otherwise we should be fine. DDoSing our server is quite easy anyway, because we have many LAPIS/SILO instances that do heavy computations.

We leave this endpoint open.

File: website/src/middleware/loggingMiddleware.ts (hunk 5 above)
Comment (fengelniederhammer) [reply website/src/middleware/loggingMiddleware.ts:9]: As discussed: For now we'll leave it on `info`, until we notice that it's too much.
